tests/README.md

```diff
@@ -113,51 +113,49 @@ snip_config_50_v3 # Step 3: switch 50% traffic to v3
 ```
 
 For commands that produce output, pass the snip and expected output to an appropriate
-`_run_and_verify_` function. For example:
+`_verify_` function. For example:
 
 ```sh
-_run_and_verify_same snip_set_up_the_cluster_3 "$snip_set_up_the_cluster_3_out"
+_verify_same snip_set_up_the_cluster_3 "$snip_set_up_the_cluster_3_out"
 ```
 
-For situations where you need to perform more than one verify check, you can
-run the snip and capture the command output in a variable, and then compare
-it to the expected output:
+The verify functions first run the snip function and then compare the result to the
+expected output. The framework includes the following built-in verify functions:
 
-```sh
-out=$(snip_tripping_the_circuit_breaker_1 2>&1)
-_verify_contains "$out" "Code 200 :" "snip_tripping_the_circuit_breaker_1"
-_verify_contains "$out" "Code 503 :" "snip_tripping_the_circuit_breaker_1"
-```
-
-The framework includes the following built-in verify functions:
-
-1. **`_verify_same`** `out` `expected` `msg`
-
-   Verify that `out` is exactly the same as `expected`. Failure messages will include
-   the specified `msg`.
+1. **`_verify_same`** `func` `expected`
 
-1. **`_verify_contains`** `out` `expected` `msg`
+   Runs `func` and compares the output with `expected`. If they are not the same,
+   exponentially back off and try again, 5 times by default. The number of retries
+   can be changed by setting the `VERIFY_RETRIES` environment variable.
 
-   Verify that `out` contains the substring `expected`. Failure messages will include
-   the specified `msg`.
+1. **`_verify_contains`** `func` `expected`
 
-1. **`_verify_not_contains`** `out` `expected` `msg`
+   Runs `func` and compares the output with `expected`. If the output does not
+   contain the substring `expected`, exponentially back off and try again, 5 times
+   by default. The number of retries can be changed by setting the `VERIFY_RETRIES`
+   environment variable.
 
-   Verify that `out` does not contains the substring `expected`. Failure messages will include
-   the specified `msg`.
+1. **`_verify_not_contains`** `func` `expected`
 
-1. **`_verify_first_line`** `out` `expected` `msg`
+   Runs `func` and compares the output with `expected`. If the command execution fails
```

Comment:
This is the only command that mentions that if command execution fails, it will try again.  This should be true for all of them, so I think it would be good to mention that either in each command or elsewhere once as general information.

Reply:
The reason I don't think it's necessary in the other functions is because if they fail to execute the output will NOT be as expected (because they are looking for something specific in the output), so they will not return success. verify_not_contains was the only tricky one that needed to check for failure, otherwise it was returning on failed commands with a successful compare (string is missing from output, as expected). Maybe mentioning command failure is just confusing things, and I shouldn't mention it at all?

Reply:
Now I understand your reasoning.  However, by calling out the behavior on command failure for just this case, it reads to me like it's implying different behavior in those other cases.  My two cents, anyway.  :-)

Reply:
That makes sense. I'll update the doc to make it clear that command failure -> retry in all verify functions.

Reply:
Turns out that we do need a different behavior for the other cases. There are examples where the actual command does fail but the expected output is what we're looking for from the snip. So, _verify_not_contains is the only one that we don't want to fail verify if the command fails, because that makes no sense.